### DISCUSSION TITLE
Is `a + yield 1;` valid?

### DISCUSSION THREAD

**Discussion by Josh-Cena** [babel/babel#14858] [Q&A] [answered]
Was playing with parsers, and discovered [this throws](https://babeljs.io/repl#?browsers=chrome%2040&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=IYAg1CCeCWCmA2ATEBGA3AKCA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=&prettier=false&targets=&version=7.18.12&externalPlugins=&assumptions=%7B%7D):

```js
a + yield 1; // Missing semicolon.
```

I already have `sourceType` set to `script` so this should be in sloppy mode. Should this be valid?

(I'd admit I didn't look into the parameterized productions on this; would appreciate some guidance)

**Comment by nicolo-ribaudo** [comment, marked as answer]
It's not valid: on the RHS of [AdditiveExpression](https://tc39.es/ecma262/#prod-AdditiveExpression) there is MultipliativeExpression (`*`, `/`, `%`), which recursively includes the various binary operator and the unary operators included in [UnaryExpression](https://tc39.es/ecma262/#prod-UnaryExpression) (including `await 1`).

`yield` is different: it's defined in [AssignmentExpression](https://tc39.es/ecma262/#prod-AssignmentExpression), so it's not valid as the RHS of an additive expression (`a + yield 1` would be like `a + y=1`).

**Reply by Josh-Cena** [reply]
Why can't it be `(a + yield); 1;` via ASI?

**Reply by Josh-Cena** [reply]
Ah, I see the [note](https://tc39.es/ecma262/#sec-identifiers) here:

> yield and await are permitted as [BindingIdentifier](https://tc39.es/ecma262/#prod-BindingIdentifier) in the grammar, and prohibited with [static semantics](https://tc39.es/ecma262/#sec-static-semantic-rules) below, to prohibit automatic semicolon insertion in cases such as
>
> ```js
> let
> await 0;
> ```

Thanks anyway @nicolo-ribaudo!

**Reply by nicolo-ribaudo** [reply]
Because ASI only inserts semicolons when there is a newline, not inside a single line.

Note that my answer above (https://github.com/babel/babel/discussions/14858#discussioncomment-3404599) was assuming that `a + yield 1` is inside a generator.
- Inside a generator, `(a + yield)` is invalid for the same reason
- Outside of a generator, `a + yield 1` is invalid for the same reason that `a + b 1` is invalid (the missing semicolon)

**Reply by Josh-Cena** [reply]
Oh oops, I forgot ASI needs line breaks😅 Indeed, adding a line break between `yield` and `1` works. Thanks!
